fish-rust/src/curses.rs

```diff
@@ -0,0 +1,304 @@
+            Ok(idx) => Some(&term.overrides[idx].1),
+            Err(idx) => {
+                let mut result = vec![b'\0'; 100];
+                unsafe {
+                    let mut area = result.as_mut_ptr() as *mut libc::c_char;
+                    let area = std::ptr::addr_of_mut!(area);
+                    if ffi::tgetstr(id.as_ptr(), area).is_null() {
+                        return None;
+                    }
+                }
+                term.overrides
+                    .insert(idx, (*self, String::from_utf8(result).unwrap()));
+                Some(&term.overrides[idx].1)
+            }
+        }
+    }
+}
+
+impl<'a> Capability<'a> for Number {
+    type Result = Option<i32>;
+
+    fn lookup(&self, _: &'a mut Term) -> Self::Result {
+        unsafe {
+            match tgetnum(self.0.as_ptr()) {
+                -1 => None,
+                n => Some(n),
+            }
+        }
+    }
+}
+
+impl<'a> Capability<'a> for Flag {
+    type Result = bool;
+
+    fn lookup(&self, _: &'a mut Term) -> Self::Result {
+        unsafe { tgetflag(self.0.as_ptr()) }
+    }
+}
+
+/// Calls the curses `setupterm()` function with the provided `$TERM` value `term` (or a null
+/// pointer in case `term` is null) for the file descriptor `fd`.
+///
+/// Note that the `errret` parameter is provided to the function, meaning curses will not write
+/// error output to stderr in case of failure.
+pub fn setup(term: Option<&str>, fd: i32) -> bool {
+    // If cur_term is already initialized for a different $TERM value, calling setupterm() again
+    // will leak memory. Call reset() first to free previously allocated resources.
+    unsafe {
+        reset();
+
+        let result = {
+            let mut err = 0;
+            if let Some(term) = term {
+                let term = CString::new(term).unwrap();
+                ffi::setupterm(term.as_ptr(), fd, &mut err)
+            } else {
+                ffi::setupterm(core::ptr::null(), fd, &mut err)
+            }
+        };
+
+        if result == self::OK {
+            *TERM.lock().expect("Mutex poisoned!") = Some(Term::new());
+            true
+        } else {
+            *TERM.lock().expect("Mutex poisoned!") = None;
+            false
+        }
+    }
+}
+
+/// Whether or not the curses library has been initialized.
+pub fn is_initialized() -> bool {
+    unsafe { !ffi::cur_term.is_null() }
+}
+
+/// Resets the curses `cur_term` TERMINAL pointer. Any previous term objects are invalidated!
+pub unsafe fn reset() {
+    if is_initialized() {
+        ffi::del_curterm(cur_term);
+        ffi::cur_term = core::ptr::null();
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Code {
+    /// The two-char termcap code for the capability, followed by a nul.
+    code: [u8; 3],
+}
+
+impl Code {
+    const fn new(code: &str) -> Code {
+        let code = code.as_bytes();
+        Code {
+            code: [code[0], code[1], b'\0'],
```

Comment:
It's a `const` fn so a panic here could be a compile-time error. The two-byte limit is hardcoded into termcap, it's not from me.

Reply:
My concern really is just that someone tries to construct a Code with the wrong string - like the probably better known terminfo name instead of the termcap one.

A compile-time error would do just fine. I'd assume that only works for literals? (which would be fine)

Reply:
I removed it from the latest update by not caching ncurses-returned strings at all.

Reply:
Yeah, if it's not a literal string then it wouldn't be a compile-time error but the type is also private to the module and not directly exposed so I think the panic I added per your request should cover it.